### PR TITLE
Mongoose - improve Schema type information

### PR DESCRIPTION
**Pull request by frogcjn**
1. Add details on `pre` and `post` on Schema
   Mongoose use `hooks-fixed` package to add hook methods, and should be added some detailed callback information.
   See https://github.com/vkarpov15/hooks-fixed
2. Add `static` and `methods` on Schema
   See http://mongoosejs.com/docs/2.7.x/docs/methods-statics.html

### CODE REVIEW

File: mongoose/mongoose.d.ts [hunk 1]
```diff
@@ -269,14 +273,42 @@ declare module "mongoose" {
...
-    pre(method: string, callback: Function): Schema;
+
+    pre(method: string, fn: HookSyncCallback): Schema;
+    pre(method: string, fn: HookSyncCallback, errorCb: HookErrorCallback): Schema;
+    pre(method: string, isAsync: boolean, fn: HookAsyncCallback): Schema;
+    pre(method: string, isAsync: boolean, fn: HookAsyncCallback, errorCb: HookErrorCallback): Schema;
+    post(method: string, fn: HookSyncCallback): Schema;
+    post(method: string, fn: HookSyncCallback, errorCb: HookErrorCallback): Schema;
+    post(method: string, isAsync: boolean, fn: HookAsyncCallback): Schema;
+    post(method: string, isAsync: boolean, fn: HookAsyncCallback, errorCb: HookErrorCallback): Schema;
```
Comment (mhegazy) [inline mongoose/mongoose.d.ts]: please merge the last two overloads:

```ts
post(method: string, isAsync: boolean, fn: HookAsyncCallback, errorCb?: HookErrorCallback): Schema;
``

File: mongoose/mongoose.d.ts [hunk 2]
```diff
@@ -269,14 +273,42 @@ declare module "mongoose" {
...
+
+  export interface HookAsyncCallback {
+      (next: Function, done: Function, ...hookArgs:any[]): any;
+  }
+
+  export interface HookErrorCallback {
+      (error: Error): any;
+  }
+
+  export interface HookNextFunction {
```
Comment (mhegazy) [inline mongoose/mongoose.d.ts]: this does not seem to be used anywhere. what is it used for?

Comment (frogcjn) [inline mongoose/mongoose.d.ts]: They are used in `pre(method: string, fn: HookSyncCallback, errorCb?: HookErrorCallback): Schema;`

Comment (frogcjn) [inline mongoose/mongoose.d.ts]: Fixed

File: mongoose/mongoose.d.ts [hunk 3]
```diff
@@ -269,14 +273,43 @@ declare module "mongoose" {
...
+      (next: HookNextFunction, done: HookDoneFunction, ...hookArgs:any[]): any;
+  }
+
+  export interface HookErrorCallback {
+      (error: Error): any;
+  }
+
+  export interface HookNextFunction {
+      (...hookArgs:any[]): any;
+      (error: Error): any;
```
Comment (mhegazy) [inline mongoose/mongoose.d.ts]: The second argument is not needed. the first one subsumes it. if you want to keep them both, then make the less specific one first.

Comment (frogcjn) [inline mongoose/mongoose.d.ts]: @mhegazy fixed

**Pull request merged**